lib/grandpa/tracker_votes.go

```diff
@@ -0,0 +1,179 @@
+	if !has {
+		// add new block hash in tracker
+		vt.cleanup()
+		elementData := newBlockHashAuthID(blockHash, authorityID)
+		element := vt.linkedList.PushFront(elementData)
+		data := voteMessageMapData{
+			peerID:  peerID,
+			message: voteMessage,
+			element: element,
+		}
+		vt.mapping[blockHash] = authorityIDToData{
+			authorityID: data,
+		}
+		return
+	}
+
+	data, voteExists := voteMessages[authorityID]
+	if voteExists {
+		// vote already exists so override the vote for the authority ID;
+		// do not move the list element in the linked list to avoid
+		// someone re-sending an equivocatory vote message and going at the
+		// front of the list, hence erasing other possible valid vote messages
+		// in the tracker.
+		data.peerID = peerID
+		data.message = voteMessage
+		voteMessages[authorityID] = data
+		return
+	}
+
+	// Add new authority ID in existing block hash map
+	vt.cleanup()
+	elementData := newBlockHashAuthID(blockHash, authorityID)
+	element := vt.linkedList.PushFront(elementData)
+	data = voteMessageMapData{
+		peerID:  peerID,
+		message: voteMessage,
+		element: element,
+	}
+	voteMessages[authorityID] = data
+}
+
+// cleanup removes the oldest vote message from the tracker
+// if the number of vote messages is at the tracker capacity.
+// This method is designed to be called automatically from the
+// add method and should not be called elsewhere.
+func (vt *votesTracker) cleanup() {
+	if vt.linkedList.Len() < vt.capacity {
+		return
+	}
+
+	oldestElement := vt.linkedList.Back()
+	vt.linkedList.Remove(oldestElement)
+
+	oldestData := oldestElement.Value.(blockHashAuthID)
+	authIDToData := vt.mapping[oldestData.blockHash]
+
+	delete(authIDToData, oldestData.authorityID)
+	if len(authIDToData) == 0 {
+		delete(vt.mapping, oldestData.blockHash)
+	}
+}
+
+// delete deletes all the vote messages for a particular
+// block hash from the vote messages tracker.
+func (vt *votesTracker) delete(blockHash common.Hash) {
+	authIDToData, has := vt.mapping[blockHash]
+	if !has {
+		return
+	}
+
+	for _, data := range authIDToData {
+		vt.linkedList.Remove(data.element)
+	}
+
+	delete(vt.mapping, blockHash)
+}
+
+// getMessagesForBlockHash returns all the vote messages
+// for a particular block hash from the tracker as a slice
+// of networkVoteMessage.
+// It returns nil if the block hash does not exist.
+func (vt *votesTracker) getMessagesForBlockHash(
+	blockHash common.Hash) (messages []networkVoteMessage) {
+	authIDToData, ok := vt.mapping[blockHash]
+	if !ok {
+		// Note authIDToData cannot be empty
+		return nil
+	}
+
+	messages = make([]networkVoteMessage, 0, len(authIDToData))
+	for _, data := range authIDToData {
+		message := networkVoteMessage{
+			from: data.peerID,
+			msg:  data.message,
+		}
+		messages = append(messages, message)
+	}
+	return messages
+}
+
+// forEach runs the function `f` on each
+// peer id + message stored in the tracker.
+func (vt *votesTracker) forEach(
```

Comment:
nice!

Reply:
For the sake of knowledge and discussion, here's a discussion on the similar `forEach` method for the commits tracker:

https://github.com/ChainSafe/gossamer/pull/2490#discussion_r856406631